filesystems/drs/src/main/scala/cromwell/filesystems/drs/DrsResolver.scala

```diff
@@ -14,40 +14,69 @@ object DrsResolver {
 
   private val GcsProtocolLength: Int = 5 // length of 'gs://'
 
-  private def resolveError[A](drsPath: DrsPath)(throwable: Throwable): IO[A] = {
+  private def resolveError[A](pathAsString: String)(throwable: Throwable): IO[A] = {
     IO.raiseError(
       new RuntimeException(
-        s"Error while resolving DRS path: $drsPath. Error: ${ExceptionUtils.getMessage(throwable)}"
+        s"Error while resolving DRS path: $pathAsString. Error: ${ExceptionUtils.getMessage(throwable)}"
       )
     )
   }
 
-  private def getFileNameAndGsUri(drsPath: DrsPath): IO[(Option[String], Option[String])] = {
+  private def getDrsPathResolver(drsPath: DrsPath): IO[DrsPathResolver] = {
     val drsFileSystemProviderOption = drsPath.drsPath.getFileSystem.provider.cast[DrsCloudNioFileSystemProvider]
 
     val noFileSystemForDrsError = s"Unable to cast file system provider to DrsCloudNioFileSystemProvider for DRS path $drsPath."
-    val fields = NonEmptyList.of(MarthaField.FileName, MarthaField.GsUri)
 
     for {
       drsFileSystemProvider <- toIO(drsFileSystemProviderOption, noFileSystemForDrsError)
-      marthaResponse <- drsFileSystemProvider.drsPathResolver.resolveDrsThroughMartha(drsPath.pathAsString, fields)
-    } yield (marthaResponse.fileName, marthaResponse.gsUri)
+    } yield drsFileSystemProvider.drsPathResolver
   }
 
-  def getGsUri(drsPath: DrsPath): IO[String] = {
+  private def getGsUriFileNameBondProvider(pathAsString: String,
+                                           drsPathResolver: DrsPathResolver
+                                          ): IO[(Option[String], Option[String], Option[String])] = {
+    val fields = NonEmptyList.of(MarthaField.GsUri, MarthaField.FileName, MarthaField.BondProvider)
+    for {
+      marthaResponse <- drsPathResolver.resolveDrsThroughMartha(pathAsString, fields)
+    } yield (marthaResponse.gsUri, marthaResponse.fileName, marthaResponse.bondProvider)
+  }
+
+  /** Returns the `gsUri` if it ends in the `fileName` and the `bondProvider` is empty. */
+  def getSimpleGsUri(pathAsString: String,
+                     drsPathResolver: DrsPathResolver): IO[Option[String]] = {
     val gsUriIO = for {
-      fileNameAndGsUri <- getFileNameAndGsUri(drsPath)
-      (_, gsUriOption) = fileNameAndGsUri
-      gsUri <- IO.fromEither(gsUriOption.toRight(UrlNotFoundException(GcsScheme)))
-    } yield gsUri
+      tuple <- getGsUriFileNameBondProvider(pathAsString, drsPathResolver)
+      (gsUriOption, fileNameOption, bondProviderOption) = tuple
+    } yield {
+      for {
+        // Only return gsUri that do not use Bond
+        gsUri <- if (bondProviderOption.isEmpty) gsUriOption else None
+        if {
+          // Do not return the gsUri if it does not end in /fileName
+          fileNameOption match {
+            case Some(fileName) if !gsUri.endsWith(s"/$fileName") => false
+            case _ => true
+          }
+        }
+      } yield gsUri
```

Comment:
Rather than putting a forcomp in the yield of another forcomp might this be broken out into a function?

Reply:
pulled out into yet-another-overloaded function